### PR TITLE
fix: fix predictions and output for multline csoaa ldf

### LINKED ISSUE #4007: csoaa_ldf predicts an extra label for "shared"
### Describe the bug

Newer versions of VW predict an additional label for the "shared" features in the csoaa_ldf mode. Also, the loss is suspicious...

### How to reproduce

Dataset ("csoaa_ldf_example.txt"):

```
shared |u real auto
1:0 |a Real_f Auto_f RealAuto_f
2:1 |a Toy_f  Auto_f ToyAuto_f
3:1 |a Real_f Air_f  RealAir_f
4:1 |a Toy_f  Air_f  ToyAir_f

shared |u toy auto
1:1 |a Real_f Auto_f RealAuto_f
2:0 |a Toy_f  Auto_f ToyAuto_f
3:1 |a Real_f Air_f  RealAir_f
4:1 |a Toy_f  Air_f  ToyAir_f

shared |u real air
1:1 |a Real_f Auto_f RealAuto_f
2:1 |a Toy_f  Auto_f ToyAuto_f
3:0 |a Real_f Air_f  RealAir_f
4:1 |a Toy_f  Air_f  ToyAir_f

shared |u toy air
1:1 |a Real_f Auto_f RealAuto_f
2:1 |a Toy_f  Auto_f ToyAuto_f
3:1 |a Real_f Air_f  RealAir_f
4:0 |a Toy_f  Air_f  ToyAir_f
```

Old VW behavior (which looks reasonable):
```
$ vw --version
8.6.1
$ vw --csoaa_ldf=m -d csoaa_ldf_example.txt -qua -p predict_old.txt
creating quadratic features for pairs: ua
predictions = predict_old.txt
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = csoaa_ldf_example.txt
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.000000 0.000000            1            1.0    known        1       16
0.500000 1.000000            2            2.0    known        1       16
0.500000 0.500000            4            4.0    known        0       16

finished run
number of examples = 4
weighted example sum = 4.000000
weighted label sum = 0.000000
average loss = 0.750000
total feature number = 64
$ cat predict_old.txt
1
0
0
0

1
0
0
0

0
2
0
0

0
0
3
0
```

New VW behavior:

```
$ ./../vowpal_wabbit/build/vowpalwabbit/cli/vw --version
9.1.0 (git commit: a78e4221d)
$ ./../vowpal_wabbit/build/vowpalwabbit/cli/vw --csoaa_ldf=m -d csoaa_ldf_example.txt -qua -p predict_new.txt
creating quadratic features for pairs: ua
predictions = predict_new.txt
using no cache
Reading datafile = csoaa_ldf_example.txt
num sources = 1
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
Enabled reductions: gd, scorer-identity, csoaa_ldf, shared_feature_merger
Input label = cs
Output pred = multiclass
average  since         example        example        current        current  current
loss     last          counter         weight          label        predict features
0.000000 0.000000            1            1.0          known              1       48
0.000000 0.000000            2            2.0          known              1       48
-1701411 -3402823            4            4.0          known              0       48

finished run
number of examples = 4
weighted example sum = 4.000000
weighted label sum = 0.000000
average loss = -170141173319264429905852091742258462720.000000
total feature number = 192
$ cat predict_new.txt
1
0
0
0
0

1
0
0
0
0

0
0
2
0
0

0
0
0
3
0
```

Note:
1. Suspicious loss value
2. 5 predicted labels per examples, rather than expected 4.

### Version

9.1.0 (git commit: a78e4221d)

### OS

WSL2

### Language

CLI

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jackgerrits**
Resolves #4007

### CODE REVIEW

**Review by bassmang — APPROVED** [review]

File: vowpalwabbit/core/src/reductions/csoaa_ldf.cc [hunk 1]
```diff
@@ -606,7 +609,17 @@ void finish_multiline_example(VW::workspace& all, ldf& data, VW::multi_ex& ec_se
 {
   if (!ec_seq.empty())
   {
+    if (ec_seq.size() > 1 && COST_SENSITIVE::ec_is_example_header(*ec_seq[0]))
+    {
+      std::swap(ec_seq[0]->pred, ec_seq[1]->pred);
+      std::swap(ec_seq[0]->tag, ec_seq[1]->tag);
+    }
     output_example_seq(all, data, ec_seq);
```
Comment (lalo) [inline vowpalwabbit/core/src/reductions/csoaa_ldf.cc:617]: why not handle shared inside output_?

Comment (jackgerrits) [reply vowpalwabbit/core/src/reductions/csoaa_ldf.cc:617]: No reason in particular

**Review by lalo — APPROVED** [review]
